Update Extend_Euclid.mdx (#2969)
cpinitiative/usaco-guide: content/6_Advanced/Extend_Euclid.mdx

```cpp
ll euclid(ll a, ll b) {
	for (; b; swap(a, b)) {
		ll k = a / b;
		a -= k * b;
	}
	return a; // gcd(a , b)
}
```

</CPPSection>
<PySection>

```py
def euclid(a: int, b: int) -> int:

>     """
>     Returns the GCD of two positive, non-zero numeric values.
>     """
> 

Reviewer: ```suggestion
```

People know what the function does.

    assert int(a) > 0 and int(b) > 0, \
        "Arguments must be positive, non-zero numeric values."

>     while b > 0: # run Euclidean algo

Reviewer: ```suggestion
    while b > 0:
```

Author: Just trying to keep it close to the c++ version.  They have a comment when the algo portion begins, so that's why I put it there.

Is there a USACO Python standards reference I could use for future updates?  I feel like a lot of these changes could have been avoided if that were the case.

        k = a // b
	# subtract multiples of one equation from the other.
        a -= b * k

>         a, b = b, a  # swap a and b

Reviewer: ```suggestion
        a, b = b, a
```

}
```

</CPPSection>
<PySection>

```py
def euclid(a: int, b: int) -> int:
    """Returns integer representing GCD of two positive, non-zero numeric values.
    """

    assert int(a) > 0 and int(b) > 0, "Arguments must be positive, non-zero numeric values."
	
    def swap(x, y):
        tmp = x
        x = y
        y = tmp
        return x, y

>     # Do while 'b' value is greater than zero.

Reviewer 2: ```suggestion
```

>     def swap(x, y):
>         tmp = x
>         x = y
>         y = tmp
>         return x, y
> 
>     while b > 0: # run Euclidean algo

Reviewer: ```suggestion
    while b > 0: # run Euclidean algo
```

>         # k == floor division of 'a' over 'b.'

Reviewer 2: ```suggestion
```

	# subtract multiples of one equation from the other.
        a -= b * k

>         a, b = swap(a, b)

Reviewer: ```suggestion
        a, b = b, a
```

Author: Good suggestion! I understand how to swap without a function, but I was trying to stay close to the C++ implementation.

How many approvers are on this pull request?

Reviewer: Currently 1

}
```

</CPPSection>
<PySection>

```py
def euclid(a: int, b: int) -> int:
    """Returns integer representing GCD of two positive, non-zero numeric values.
    """

    # Validate arguments.
    assert int(a) > 0 and int(b) > 0, "Arguments must be positive, non-zero numeric values."

    # Create tuple object from 'a' and 'b.'

>     ab: list[int] = [int(a), int(b)]

Author: Easier to do an "in-place" swap in Python using a function with a list.  I can just return values, I suppose.

The code is like this because I wanted to just mimic what was already being done with the C++ version.

    def swap(ab: list[int]) -> None:
        ab[0], ab[1] = ab[1], ab[0]

>     # Do while 'b' value is greater than zero.

Reviewer 2: self explanatory comments

```

</CPPSection>
<PySection>

```py
def extend_euclid(a: int, b: int) -> list[int]:

>     """
>     Returns a list of Bézout coefficients and the
>     GCD of two positive, non-zero numeric values.
>     """
> 
>     assert int(a) > 0 and int(b) > 0, \

Reviewer: ```suggestion
    assert int(a) > 0 and int(b) > 0, \
```

        "Arguments must be positive, non-zero numeric values."

    x_arr: list[int] = [1, 0, int(a)]
    y_arr: list[int] = [0, 1, int(b)]
    q: int = -1

>     def swap(x, y):
>         tmp = x
>         x = y
>         y = tmp
>         return x, y
> 
> 	"""
>     We know that:
>         (1*a) + (0*b) == a
>         (0*a) + (1*b) == b
>     """

Reviewer: ```suggestion
```

Author: I feel like having commentary/light documentation is not a bad thing. That exact comment is on the C++ version.

```

</CPPSection>
<PySection>

```py
def extend_euclid(a: int, b: int) -> list[int]:
    """Returns integer list representing Bézout coefficients and the
    GCD of two positive, non-zero numeric values.
    """

    assert int(a) > 0 and int(b) > 0, \
        "Arguments must be positive, non-zero numeric values."

    xarr: list[int] = [1, 0, int(a)]

>     yarr: list[int] = [0, 1, int(b)]

Reviewer 2: add similar comments from the cpp representation

    q: int = -1

    def swap(x, y):
        tmp = x
        x = y
        y = tmp
	return x, y

    # Run and continue for 'b' > 0.
    while yarr[2] > 0:
        q = xarr[2] // yarr[2]

>         # Update all three values.

Reviewer 2: ```suggestion
```

```

</CPPSection>
<PySection>

```py
def extend_euclid(a: int, b: int) -> list[int]:

>     """Returns integer list representing Bézout coefficients and the
>     GCD of two positive, non-zero numeric values.
>     """

Reviewer 3: ```suggestion
    """
    Returns a list of Bézout coefficients and the
    GCD of two positive, non-zero numeric values.
    """
```

```

</CPPSection>
<PySection>

```py
def extend_euclid(a: int, b: int) -> list[int]:
    """
    Returns a list of Bézout coefficients and the
    GCD of two positive, non-zero numeric values.
    """

    assert int(a) > 0 and int(b) > 0, \
        "Arguments must be positive, non-zero numeric values."

    x_arr: list[int] = [1, 0, int(a)]
    y_arr: list[int] = [0, 1, int(b)]
    q: int = -1

>     def swap(x, y):
>         tmp = x
>         x = y
>         y = tmp
>         return x, y
> 
>     # We know that:

Reviewer: ```suggestion
    # We know that:
```

    #   (1*a) + (0*b) == a

>     #   (0*a) + (1*b) == b

Reviewer: multiline this

Author: I don't understand. Can you please clarify "multiline this"?

Reviewer: use:
```py
"""
We know that:
etcetc
etcetc
"""
```

Author: Ah, gotcha.  Thanks for the visual!

    while y_arr[2] > 0: # run extended Euclidean algo.
        q = x_arr[2] // y_arr[2]
        for i in range(3):
	    # keep subtracting multiple of one equation from the other.
            x_arr[i] -= y_arr[i]*q

>         x_arr, y_arr = swap(x_arr, y_arr)

Reviewer: ```suggestion
        x_arr, y_arr = y_arr, x_arr
```

Reviewer: do this.

```

</CPPSection>
<PySection>

```py
def extend_euclid(a: int, b: int) -> list[int]:
    """
    Returns a list of Bézout coefficients and the
    GCD of two positive, non-zero numeric values.
    """

    assert int(a) > 0 and int(b) > 0, \
        "Arguments must be positive, non-zero numeric values."

    x_arr: list[int] = [1, 0, int(a)]
    y_arr: list[int] = [0, 1, int(b)]
    q: int = -1

    def swap(x, y):
        tmp = x
        x = y
        y = tmp
        return x, y

	"""
    We know that:
        (1*a) + (0*b) == a
        (0*a) + (1*b) == b
    """
    while y_arr[2] > 0: # run extended Euclidean algo.
        q = x_arr[2] // y_arr[2]
        for i in range(3):

> 	    # keep subtracting multiple of one equation from the other.

Reviewer 3: tabs to indent

        q = x_arr[2] // y_arr[2]
        for i in range(3):
	    # keep subtracting multiple of one equation from the other.

>             x_arr[i] -= y_arr[i]*q

Reviewer 3: wait why remove the spaces

```

</CPPSection>
<PySection>

```py

def euclid(a: int, b: int) -> int:
    """Recursive Euclidean GCD."""
    if not b:
        return a

>     return euclid(b, a % b)

Reviewer 3: a ternary would work too (optional)

Author: Do you mean something like:

```py
return a if b else euclid(b, a % b)
```

That would work, but I think keeping it slightly more verbose might be better for a tutorial/reference website.

Reviewer 3: fair enough

```

</CPPSection>
<PySection>

```py

def euclid(a: int, b: int) -> int:
    """Recursive Euclidean GCD."""
    if not b:
        return a

>     return euclid(b, a%b)

Reviewer: ```suggestion
    return euclid(b, a % b)
```

Reviewer: if sans or envy told you to do otherwise, ignore this

Author: See: https://peps.python.org/pep-0008/#other-recommendations

Reviewer: https://usaco.guide/general/adding-solution?lang=py#code-conventions

```

</CPPSection>
<PySection>

```py

def extend_euclid(a: int, b: int) -> list[int]:
    if not b:
        return [1, 0]
    p = ext_euclid(b, a % b)

>     return [p[1], p[0] - (a//b) * p[1]]

Reviewer 3: spaces for operators

```

</CPPSection>
<PySection>

```py

def extend_euclid(a: int, b: int) -> list[int]:
    if not b:
        return [1, 0]

>     p = ext_euclid(b, a%b)

Reviewer: ```suggestion
    p = ext_euclid(b, a % b)
```

Reviewer: if sans or envy told you to do otherwise, ignore this

Author: See: https://peps.python.org/pep-0008/#other-recommendations

Reviewer: https://usaco.guide/general/adding-solution?lang=py#code-conventions

>     return [p[1], p[0] - (a//b)*p[1]]

Reviewer: ```suggestion
    return [p[1], p[0] - (a // b) * p[1]]
```

Reviewer: if sans or envy told you to do otherwise, ignore this

Author: See: https://peps.python.org/pep-0008/#other-recommendations

Reviewer: https://usaco.guide/general/adding-solution?lang=py#code-conventions

we don't follow pip strictly, although if sans or envy say otherwise, then its okay.

Reviewer 4: the original code is fine as is

```

</CPPSection>
<PySection>

```py
from math import gcd

def inv_general(a: int, b: int) -> int:
    """Returns modular inverse of two positive, non-zero integer values."""
    arr = ext_euclid(a, b)
    assert arr[2] == 1, "GCD must be 1."
    return arr[0] + (arr[0] < 0) * b

def test_inv_general():
    for b in range(1, 101):
        for a in range(101):
            if gcd(a, b) == 1:
                x = inv_general(a, b)
                assert (a * x - 1) % b == 0

>                 assert 0 <= x and x <  b

Reviewer 3: i'd put this in a `if __name__ == "__main__"` or somth
